### PR TITLE
[prometheus-thanos] Fix compact and store-gateway templates breaking when there are no extra volumes defined

### LINKED ISSUE #296: prometheus-thanos: compact produces broken yaml when no custom volumes are defined
**Version of Helm and Kubernetes**:

```
Client Version: v1.17.2
Server Version: v1.14.9-eks-c0eccc
```

**Which chart in which version**:

`prometheus-thanos` - tested with versions `2.0.0` through `3.0.0`

**What happened**:

When the user doesn't enable `compact.persistentVolume` or configure any volumes, the yaml generated for compact-statefulset is invalid and looks something like: 

```yaml
...
  template:
    metadata:
      ...
    spec:
      containers:
        - name: prometheus-thanos-compact
          imagePullPolicy: IfNotPresent
          image: ...
          args:
          - compat
            ...
          ports:
          - name: monitoring
            containerPort: 10902
          resources:
            ...
          volumeMounts:
            - mountPath: /data
              name: storage-volume
        - name: storage-volume
          emptyDir: {}
```

The same issue can show itself in various ways depending on the configuration options set. In my case, when I was setting `compact.nodeSelector`, helm was returning the following error:

```
YAML parse error on prometheus-thanos/templates/compact-statefulset.yaml: error converting YAML to JSON: yaml: line 62: did not find expected key
```
Inconsistent indentation in `templates/compact-statefulset.yaml` made it challenging to figure out what was the actual problem.

### PR DESCRIPTION
**Pull request by rozcietrzewiacz**
#### What this PR does / why we need it:

The prometheus-thanos chart was breaking (producing invalid yaml) when installed with disabled `compact.persistentVolume` and no custom volumes (both are defaults).

#### Which issue this PR fixes
  - fixes #296
  - most probably also #255 

#### Special notes for your reviewer:

In `compact-statefulset.yaml`, I have:
- fixed indentation
- fixed the logic bug (see details in respective commit message)
- (**moving to separate PR**) improved readability by extracting `{{ $values := .Values.compact }}` at the top of template

I also noticed the same logic issue exists in `store-gateway-statefulset.yaml`, so I ported all the above fixes there as well.

#### Checklist

- [x] [DCO](https://developercertificate.org) signed
- [x] Chart Version bumped (if the pr is an update to an existing chart)
- [x] Title of the PR starts with chart name (e.g. `[fluentd-elasticsearch]`)

### CODE REVIEW

File: charts/prometheus-thanos/templates/compact-statefulset.yaml [hunk 1]
```diff
@@ -11,7 +12,7 @@ metadata:
 spec:
   replicas: 1
   updateStrategy:
-    type: {{ .Values.compact.updateStrategy }}
+    type: {{ $values.updateStrategy }}
```
Comment (monotek) [inline charts/prometheus-thanos/templates/compact-statefulset.yaml]: why is this changed? 
we like to configure this separately for all components.

Comment (rozcietrzewiacz) [reply charts/prometheus-thanos/templates/compact-statefulset.yaml]: The functionality is not changed. Look at line 2 above: `$values` is the local variable I introduced for improved readability. It equals `.Values.compact`.

This also helps comparing templates across components (e.g. diff `compact-statefulset.yaml` and `store-gateway-statefulset.yaml`.

Comment (monotek) [reply charts/prometheus-thanos/templates/compact-statefulset.yaml]: OK, but we should keep this consistent in the whole chart.

**Comment by rozcietrzewiacz** [comment]
@monotek You're right, adding this in just two templates introduces an inconsistent standard. I will split it off as part of chart refactoring in a separate PR. Also, your first comment made me realize that `$values` is a bit unfortunate variable name choice, which can lead to confusion.

**Comment by rozcietrzewiacz** [comment]
FYI: I just found the same logic flaw with volumes in `ruler-statefulset.yaml` as well. I'll include it in this PR as well

**Comment by rozcietrzewiacz** [comment]
I updated the PR with the same volume fix added to ruler-statefulset as well.

There is still one edge case that doesn't result in a slightly unclean yaml, though acceptable to kubernetes: If you do use  `volumeClaimTemplates`, but have no additional volumes defined, you'd get an orphaned `volumes:` key in the pod template of statefulsets: 

```yaml

apiVersion: apps/v1
kind: StatefulSet
metadata:
  ...
spec:
  ...
  template:
    ...
    spec:
      containers:
        - name: ...
          ...
          volumeMounts:
            - mountPath: /data
              name: storage-volume
      ...
      volumes:
  volumeClaimTemplates:
    - metadata:
        name: storage-volume
      spec:
        ...
```

Even though this does not look great, it does not break the syntax. And avoiding it would require an extra combined condition added above `volumes:`, something like:

```yaml
      {{- if (not .Values.storeGateway.persistentVolume.enabled) or (.Values.storeGateway.volumes) }}
      volumes:
        {{- toYaml .Values.storeGateway.volumes | nindent 8 }}
        {{- if not .Values.storeGateway.persistentVolume.enabled }}
        - name: storage-volume
          emptyDir: {}
        {{- end }}
      {{- end }}
```

:thinking:  Looks a bit of an overkill to me. What do you guys think?

**Comment by rozcietrzewiacz** [comment]
@monotek @axdotl Sorry to bump this, but it seems that Rico is away from github since early December. Can someone else have a look? I've reduced the scope to the logic issues that were causing invalid yaml to be generated.
